File: lib/plugins/aws/invokeLocal/invoke.py
import argparse
import json
import logging
import sys
from time import time
from importlib import import_module

class FakeLambdaContext(object):
    def __init__(self, name='Fake', version='LATEST', timeout=6, **kwargs):
        self.name = name
        self.version = version
        self.created = time()
        self.timeout = timeout
        for key, value in kwargs.items():
            setattr(self, key, value)

    def get_remaining_time_in_millis(self):
        return int(max((self.timeout * 1000) - (int(round(time() * 1000)) - int(round(self.created * 1000))), 0))

    @property
    def function_name(self):
        return self.name

    @property
    def function_version(self):
        return self.version

    @property
    def invoked_function_arn(self):
        return 'arn:aws:lambda:serverless:' + self.name

    @property
    def memory_limit_in_mb(self):
        return 1024

    @property
    def aws_request_id(self):
        return '1234567890'


logging.basicConfig()

parser = argparse.ArgumentParser(
    prog='invoke',
    description='Runs a Lambda entry point (handler) with an optional event',
)

parser.add_argument('handler_path',
                    help=('Path to the module containing the handler function,'
                          ' omitting ".py". IE: "path/to/module"'))

parser.add_argument('handler_name', help='Name of the handler function')

if __name__ == '__main__':
    args = parser.parse_args()

    # this is needed because you need to import from where you've executed sls
    sys.path.append('.')

    module = import_module(args.handler_path.replace('/', '.'))
    handler = getattr(module, args.handler_name)

    input = json.load(sys.stdin)
    if sys.platform != 'win32':
        sys.stdin = open('/dev/tty')
    context = FakeLambdaContext(**input.get('context', {}))
    result = handler(input['event'], context)
    sys.stdout.write(json.dumps(result, indent=4))

File: lib/plugins/aws/invokeLocal/invoke.rb
require 'json'

class FakeLambdaContext
  attr_reader :function_name, :function_version

  def initialize(name: 'Fake', version: 'LATEST', timeout: 6, **options)
    @function_name = name
    @function_version = version
    @created_time = Time.now()
    @timeout = timeout
    options.each {|k,v|
      send(k, v)
    }
  end

  def get_remaining_time_in_millis
    [@timeout*1000 - ((Time.now() - @created_time)*1000).round, 0].max
  end

  def invoked_function_arn
    "arn:aws:lambda:serverless:#{function_name}"
  end

  def memory_limit_in_mb
    return '1024'
  end

  def aws_request_id
    return '1234567890'
  end

  def log_group_name
    return "/aws/lambda/#{function_name}"
  end

  def log_stream_name
    return Time.now.strftime('%Y/%m/%d') +'/[$' + function_version + ']58419525dade4d17a495dceeeed44708'
  end

  def log(message)
    puts message
  end
end


def attach_tty
  unless Gem.win_platform? || $stdin.tty? || !File.exist?("/dev/tty")
    $stdin.reopen "/dev/tty", "a+"
  end
rescue
  puts "tty unavailable"
end

if __FILE__ == $0
  unless ARGV[0] && ARGV[1]
    puts "Usage: invoke.rb <handler_path> <handler_name>"
    exit 1
  end

  handler_path = ARGV[0]
  handler_name = ARGV[1]

  input = JSON.load($stdin) || {}

  require("./#{handler_path}")

  # handler name is either a global method or a static method in a class
  # my_method or MyModule::MyClass.my_method
  handler_method, handler_class = handler_name.split(".").reverse
  handler_class ||= "Kernel"

  attach_tty

  context = FakeLambdaContext.new(**input.fetch('context', {}))
  result = Object.const_get(handler_class).send(handler_method, event: input['event'], context: context)

  puts result.to_json
end
